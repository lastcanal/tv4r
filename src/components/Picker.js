import React, { useMemo, useRef } from 'react'
import PropTypes from 'prop-types'
import { makeStyles, useTheme } from '@material-ui/core/styles'
import CreatableSelect from 'react-select/creatable'
import { connect } from 'react-redux'
import Option from './Option'
import { muiThemeToRSTheme } from '../helpers'
import {
  enableKeyboardControls,
  disableKeyboardControls,
  loadSubreddit,
} from '../actions'

const useStyles = makeStyles(({ palette, spacing }) => ({
  container: {
    margin: spacing(2),
    marginLeft: 0,
    marginRight: spacing(1),
    display: 'block',
    backgroundColor: palette.background.default,
    maxHeight: 200,
    flex: 1,
  },
}))

const makeSelectStyles = ({ palette }) => ({
  container: styles => ({
    ...styles,
    backgroundColor: palette.background.default,
    color: palette.text.contrastText,
  }),
  menu: styles => ({
    ...styles,
    backgroundColor: palette.background.paper,
    border: `4px solid ${palette.background.paper}`,
  }),
})

const nameToOption = (name) => ({
  value: (name || '').toLowerCase(), label: name,
})

const Picker = ({ value, options, dispatch }) => {
  const classes = useStyles()
  const theme = useTheme()
  const ref = useRef()

  const changeSubreddit = ({ value }) => {
    dispatch(loadSubreddit(value))
    void ref.current?.blur()
  }

  const mappedOptions = useMemo(() => (
    options.map(nameToOption)
  ), [options])

  const mappedValue = useMemo(() => (
    value ? nameToOption(value) : ''
  ), [value])

  const selectTheme = useMemo(() => (
    muiThemeToRSTheme(theme)
  ), [theme])

  const selectStyles = useMemo(() => (
    makeSelectStyles(theme)
  ), [theme])

  const onMenuOpen = () => {
    dispatch(disableKeyboardControls())
  }

  const onMenuClose = () => {
    dispatch(enableKeyboardControls())
  }

  return (
    <div className={classes.container}>
      <CreatableSelect
        ref={ref}
        isClearable={false}
        menuPlacement="bottom"
        defaultValue={mappedValue}
        onChange={changeSubreddit}
        options={mappedOptions}
        theme={selectTheme}
        maxMenuHeight={200}
        styles={selectStyles}
        onMenuOpen={onMenuOpen}
        onMenuClose={onMenuClose}
        components={{
          Option,
        }}
      />
    </div>
  )
}

Picker.propTypes = {
  options: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  value: PropTypes.string,
  dispatch: PropTypes.func,
}

const mapStateToProps = ({ selectedSubreddit, subreddits }) => ({
  value: selectedSubreddit,
  options: subreddits,
})

export default Picker |> connect(mapStateToProps)
