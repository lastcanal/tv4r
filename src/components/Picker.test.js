import Picker from './Picker'

describe('Picker', () => {
  it('should render Picker', () => {
    const onChange = jest.fn(() => {})
    const wrapper = shallow(
      <Picker options={['foo', 'bar']} value={'foo'} onChange={onChange} />,
    )

    expect(wrapper).toMatchSnapshot()
  })
})
