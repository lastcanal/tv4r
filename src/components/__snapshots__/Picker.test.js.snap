// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`picker should render Picker 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Connect(Picker)
    options={
      Array [
        "foo",
        "bar",
      ]
    }
    value="foo"
  />
</ContextProvider>
`;
